### ISSUE TITLE
describing a call throws an exception when argument is strict

### ISSUE THREAD

**Issue by Inspyro** [FakeItEasy/FakeItEasy#457]
Say we setup a strict fake:

``` csharp
var a = A.Fake<SomeInterface>(s=>s.Strict()); //A strict fake!
var b = A.Fake<SomeOtherInterface>(); //strictness does not matter...
```

and then we verify the strict fake was passed as an argument to another mock:

``` csharp
//a.ToString() is called for some reason, however a is a strict fake
A.CallTo(()=>b.SomeMethod(a)).MustHaveHappened()
```

FakeItEasy throws an Exception:

``` csharp
FakeItEasy.ExpectationException : Call to non configured method "ToString" of strict fake.
```

Call Stack:

``` csharp
   at FakeItEasy.FakeOptionsBuilderExtensions.<Strict>b__0<T>(IFakeObjectCall call)
   at FakeItEasy.Configuration.BuildableCallRule.Apply(IInterceptedFakeObjectCall fakeObjectCall)
   at FakeItEasy.Core.FakeManager.FakeItEasy.Core.IFakeCallProcessor.Process(IWritableFakeObjectCall fakeObjectCall)
   at Castle.DynamicProxy.AbstractInvocation.Proceed()
   at Castle.Proxies.ObjectProxy_4.ToString()
   at FakeItEasy.Expressions.ArgumentConstraints.EqualityArgumentConstraint.WriteDescription(IOutputWriter writer)
   at FakeItEasy.Expressions.ExpressionCallMatcher.AppendArgumentsListString(StringBuilder result)
   at FakeItEasy.Expressions.ExpressionCallMatcher.ToString()
   at FakeItEasy.Configuration.BuildableCallRule.WriteDescriptionOfValidCall(IOutputWriter writer)
   at FakeItEasy.Configuration.RuleBuilder.MustHaveHappened(Repeated repeatConstraint)
```

**Comment by adamralph** [comment]
@Inspyro thanks for raising this.

@FakeItEasy/owners I can reproduce this in LINQPad:

```csharp
void Main()
{
	var a = A.Fake<SomeInterface>(s=>s.Strict()); //A strict fake!
	var b = A.Fake<SomeOtherInterface>();
	A.CallTo(()=>b.SomeMethod(a)).MustHaveHappened();
}

public interface SomeInterface{}

public interface SomeOtherInterface
{
	void SomeMethod(SomeInterface a);
}
```

**Comment by blairconrad** [comment]
Tough one. We want to explain that the call to `b.SomeMethod` never happened with argument `a`. The default way to do that is using `ToString`, which will often be very helpful to the user.
For that reason I'm not terribly excited about changin the use of `ToString`. Catching the exception and then replacing the description with the name of the type might work.

Of course, FakeItEasy will still have executed `ToString` on `a` at that point, but only while trying to throw the `ExpectationException`, so the fact that we've altered the state may not be that big a deal, so long as we can get the `ExpectationException` out.

**Comment by blairconrad** [comment]
I would look out for the same problem when describing a call that _was made_ with a strict argument, but have not tried it.

**Comment by blairconrad** [comment]
@Inspyro, thanks for raising this.

`a.ToString()` is called because FakeItEasy is trying to describe the expectation `b.SomeMethod(a)` that was never met. To do that, it uses `a.ToString()` to describe `a`.

**Comment by adamralph** [comment]
Can we inspect the type and determine it is a strict fake and not call `ToString()` in that case? Or only call it if it is defined?

**Comment by blairconrad** [comment]
This would be nice to fix, but it's a bit of a corner case, and easily worked around.
I see no reason to rush it into the 2.0 release, especially if it would hold up work on that release.

**Comment by blairconrad** [comment]
@adamralph, `ToString` is always defined, I believe. We could check for strictness. But then we'd have to check for explicit configuration of `ToString`…

And I wonder how scoping affects that.

**Comment by adamralph** [comment]
Sorry, by 'defined' I meant 'configured'.

**Comment by blairconrad** [comment]
This whole issue brings up the broader question of FakeItEasy interacting with the fakes that that have been created for tests. It's not just the `ToString`, called when there's an error, that can cause problems:

```C#
[Test]
public void Should_not_mind_when_we_call_method_configured_with_a_different_strict_arg()
{
    var arg1 = A.Fake<SomeInterface>(s => s.Strict()); //A strict fake!
    var arg2 = A.Fake<SomeInterface>(s => s.Strict()); //A strict fake!

    var b = A.Fake<SomeOtherInterface>();

    A.CallTo(() => b.SomeMethod(arg1)).DoesNothing();

    b.SomeMethod(arg2);
}
```

```
FakeItEasy.ExpectationException : Call to non configured method "Equals" of strict fake.
at FakeItEasy.FakeOptionsBuilderExtensions.<Strict>b__0<T>(IFakeObjectCall call)
at FakeItEasy.Configuration.BuildableCallRule.Apply(IInterceptedFakeObjectCall fakeObjectCall)
at FakeItEasy.Core.FakeManager.FakeItEasy.Core.IFakeCallProcessor.Process(IWritableFakeObjectCall fakeObjectCall)
at Castle.DynamicProxy.AbstractInvocation.Proceed()
at Castle.Proxies.ObjectProxy.Equals(Object obj)
at System.Linq.Enumerable.All(IEnumerable`1 source, Func`2 predicate)
at FakeItEasy.Core.FakeManager.<>c__DisplayClass11.<FakeItEasy.Core.IFakeCallProcessor.Process>b__10(CallRuleMetadata rule)
at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
at System.Linq.Enumerable.First(IEnumerable`1 source)
at FakeItEasy.Core.FakeManager.FakeItEasy.Core.IFakeCallProcessor.Process(IWritableFakeObjectCall fakeObjectCall)
at Castle.DynamicProxy.AbstractInvocation.Proceed()
at FakeItEasyQuestions.StrictInteractions.Should_not_mind_when_we_call_method_configured_with_a_different_strict_arg() in StrictInteractions.cs: line 43 
```

How much responsibility should FakeItEasy have for shielding the fakes from the operations that FakeItEasy needs to perform to fulfill its duties?

**Comment by blairconrad** [comment]
Back to having to check for explicit configuration of ToString (https://github.com/FakeItEasy/FakeItEasy/issues/457#issuecomment-73693612), perhaps that's not something we'd need to do.

If it turns out that we want to describe a strict fake in an exception message, I'd be comfortable falling back to something like "a strict fake of type FakeType".

**Comment by adamralph** [comment]
Sounds fine to me. Is this important enough for a 1.x patch or do we think
it can wait for 2.0?

**Comment by blairconrad** [comment]
Me, I think this is an extremely edge case (sorry, @Inspyro) and it can wait until 2.0 or even later.

Also, I'm not sure what problem you're proposing fixing. The `ToString` when there's already been a missed expectation, or the exception we'd throw when checking an argument that happened to be a strict fake?

My answer doesn't really change either way, though. For `ToString`, at least for now the user will know that _something_ went wrong, and the stack trace looks clear to me (we were trying to `WriteDescriptionOfValidCall` and that eventually called `ToString`), so there's a workaround, or at least a means to diagnose the problem further (allow `ToString`, even temporarily).
I don't like it as a permanent fix, but for the nonce, I think it's okay.

For the `Equals`, we've long documented that the default argument matching is done [using `object.Equals`](https://github.com/FakeItEasy/FakeItEasy/wiki/Argument-Constraints#matching-values-exactly). Moreover, we say

> If the type to be checked does not provide an adequate Equals method, you may have to use the That.Matches method described in Custom Matching.

I'd argue that a strict fake does not provide an adequate Equals method!

For me, I'd rather focus our efforts on getting 2.0 out. As you know, I've no objection to quick releases, though, so if 2.0.1 or whatever comes out 3 days later fix(es) for this issue(s), that's great!

@Inspyro, I've been talking about the severity of the issue and the workaroundability and so on, but it's not a problem that _I've_ encountered. For now, have you been able to work around the problem and carry on with your work?

**Comment by Inspyro** [comment]
Hi, yes at the moment we use non-strict mocks instead of strict mocks. But as soon as we really need strict mocks, I will try to find a workaround. 

**Possible improvement:**
In general I think it would be great if you could at least improve the exception message / type for such cases, because analyzing the stack-trace can be tedious. 
So instead of *"Call to non configured method "ToString" of strict fake"* it would be great if you can write 
*"Call to non configured method "ToString" of strict fake "SomeOtherInterface a"*...

**Comment by blairconrad** [comment]
@Inspyro, I completely agree that the error message could be improved to at least include the faked type that's strict. Although probably not to the point of including the variable name, but as we wouldn't know that when constructing the message.

**Comment by adamralph** [comment]
The error message shouldn't mention `ToString()` at all. It has no relevance to the code example in https://github.com/FakeItEasy/FakeItEasy/issues/457#issuecomment-73692199.

I think all we have to do, as you suggested in https://github.com/FakeItEasy/FakeItEasy/issues/457#issuecomment-73692918 @blairconrad, is catch the exception when constructing the description and use the type name instead. Just one question about that, what is the state altering of which you speak there? Is that side effect of the attempting `ToString()` call with regard to call recording, etc.? In that case I don't think that's a problem since we're going to throw an exception straight away anyway.

**Comment by blairconrad** [comment]
@adamralph Yes. The error message should mention the method that's called by the user's code, not the `ToString` that's called by FakeItEasy itself. I think @Inspyro was just pushing for better identification of the strict fake _when it was necessary for FakeItEasy to complain about a method being called on a strict_.
Regardless, that's what I was suggesting. So in the original example, no complaints about calling `ToString` at all. But in this example:

```C#
var a = A.Fake<SomeInterface>(s=>s.Strict()); //A strict fake!
a.Method();
```

we could consider emitting something more like `Call to non configured method "Method" of strict fake of type "Fake SomeInterface"` instead of `Call to non configured method "Method" of strict fake`.

Although it may not be useful in too many cases, as most of the time it should be clear _which_ fake had the violation.

**Comment by adamralph** [comment]
OK, I think we should spin off a separate issue for the latter case. Let's keep this issue focused on solving the reported problem, which is reproduced by the code samples above.

**Comment by adamralph** [comment]
That new issue would be a feature, rather than a bug.

**Comment by lijunle** [comment]
Well, I have a feature request, besides the showing the calling method name in the exception message, could you please showing the calling arguments? That information is so important and be much useful for user to debug.

Should I open another issue?

**Comment by blairconrad** [comment]
@lijunle, yes, I'd definitely make that a separate issue, and we'd be happy to look at it.

**Comment by thomaslevesque** [comment]
Couldn't we exclude the methods from `Object` from the "strictness" rule? These methods (`ToString`, `Equals`, `GetHashCode`) are often used implicitly (e.g. in a dictionary, or when formatting the object in a string), and it can come as a surprise that they need to be configured as well. This behavior could be optional, e.g. by adding a parameter to `Strict()`:

    A.Fake<SomeInterface>(s => s.Strict(exceptObjectMethods: true))

It would have the same effect as configuring these methods with `.CallsBaseMethod`.

**Comment by blairconrad** [comment]
@thomaslevesque, allowing degrees of strictness is interesting, but I'd pursue in another issue.
(Although it's possible for users to add the `CallsBaseMethod` themselves if interested. I'm not sure how much demand there is for these partially-strict objects, so I'm not _super_ excited about the feature. (When am I?)).

I think that there's still an opportunity here to try to keep FakeItEasy from calling `ToString` and `Equals` when operating on a strict fake.
`ToString` should be easy. I'm not so sure about `Equals`, although we could catch an error from Equals and see if the object is a strict and Fake and emit a better message. And I doubt `FakeItEasy` ever needs to `GetHashCode`.

**Comment by thomaslevesque** [comment]
> I'm not sure how much demand there is for these partially-strict objects

Probably not much demand for it as a feature, but it solves the issue of describing calls with strict arguments, comparing strict arguments, etc. Although it makes sense when you think about it, having `ToString()` and `Equals` throw when they're not explicitly configured can be surprising to users (it was to me, at least).

**Comment by blairconrad** [comment]
Oh, sure. That make sense. I just think we still need a way to describe calls to even completely strict objects.

**Comment by thomaslevesque** [comment]
> I just think we still need a way to describe calls to even completely strict objects.

True. But the problem is that `ToString`, `Equals` and `GetHashCode` aren't supposed to throw, ever!

https://msdn.microsoft.com/en-us/library/system.object.tostring.aspx

> Your `ToString` override should not throw an exception. 

https://msdn.microsoft.com/en-us/library/bsc2ak47.aspx

> Implementations of `Equals` must not throw exceptions; they should always return a value. For example, if `obj` is null, the `Equals` method should return false instead of throwing an `ArgumentNullException`. 

https://msdn.microsoft.com/en-us/library/system.object.gethashcode.aspx

With strict fakes, we explicitly create objects that will violate these rules. So yes, we could catch the exception, but we shouldn't have to...

**Comment by adamralph** [comment]
AFIAK strict fakes exist to ensure that a SUT isn't calling anything it shouldn't. (IMHO it's something that should be almost never used as it leads to Gulliver's Travels Tests.)

Ensuring that a SUT doesn't call the object inherited methods seems way OTT and a ridiculous scenario, or can someone argue otherwise?

If not, I vote for allowing them to be called on a strict fake.

**Comment by thomaslevesque** [comment]
> If not, I vote for allowing them to be called on a strict fake.

I like it. And if someone really really wants to prevent calls to them (which I doubt), we can always add an option to make the fake even stricter...

Could it be considered a breaking change? It does change the observable behavior, but it seems unlikely that it would break anyone's tests.

**Comment by adamralph** [comment]
Could it actually break anyone's tests?

**Comment by thomaslevesque** [comment]
> Could it actually break anyone's tests?

Only if the test is expecting an exception because the call is not configured... as I said: unlikely.

**Comment by adamralph** [comment]
So unlikely that I would call this non-breaking.

BTW - `GetType()` is another one that shouldn't throw.

**Comment by adamralph** [comment]
Full list:

`Equals()`, `GetHashCode()`, `GetType()`, `ToString()`

**Comment by thomaslevesque** [comment]
> BTW - GetType() is another one that shouldn't throw.

It already doesn't throw, since it's not virtual ;)

**Comment by adamralph** [comment]
Doh! OK, so back to

`Equals()`, `GetHashCode()`, `ToString()`

**Comment by blairconrad** [comment]
Sadly, I feel a little differently about the `object` methods. From the day I read about `Strict`, I assumed it meant throw on _any_ method. [Our documentation](http://fakeiteasy.readthedocs.io/en/2.2.0/strict-fakes/) supports this view:

> **all calls** to unconfigured members are rejected,

(emphasis mine)

I prefer this interpretation. Maybe `ToString` is super expensive and people want to know when it's called. Ditto any of the other methods.

My preference would be to leave the behaviour as is and investigate fixing the reported problem by other means. I liked @thomaslevesque's original suggestion of an overload for `Strict` (or equivalent mechanism) that would loosen the strictness a little.

That being said, I could go along with a change in the default behaviour, but I think it should be classified as breaking. Someone might be using strict Fakes in a way _that conforms exactly to the documentation_, and all of a sudden their Fakes act differently.

Maybe they're testing some caching mechanism, and expect one of the object methods to be called only once so they made a strict Fake and then overrode the method for one invocation only.

**Comment by adamralph** [comment]
@blairconrad fair enough. Although I consider those to be edge cases, they are valid nonetheless.

For the purposes of this bug, I think we should leave the strictness as-is, and just fix the problem at hand.

**Comment by thomaslevesque** [comment]
I think I have an idea for solving this. We could introduce an internal interface like this:
```
internal interface IFakeObject
{
    bool Equals(object other);
    int GetHashCode();
    string ToString();
}
```
and make the fakes implement this interface explicitly. When we need to call one of these methods on a fake, we would just cast it to `IFakeObject` and make the call via the interface. The default implementation for non-strict fakes would call the explicit implementation. And the `Strict` rule would need to exclude calls made via that interface, of course.

Now, the problem is that it's not easy to find all places in the code where we can call one of these methods on a fake...

**Comment by thomaslevesque** [comment]
I POCed this solution and it seems to work. The only issue is that `IFakeObject` needs to be public, otherwise Castle can't generate the proxy (not sure why, since we have an `InternalsVisibleTo` that should allow DynamicProxyGenAssembly2 to access it). This is actually the same issue as `ITaggable`.

I'll send a PR

**Comment by blairconrad** [comment]
Sounds like very good work, @thomaslevesque. I'm sad about the visibility, though. Do you think it's worthwhile bothering a Castle.Core expert who might have insight? Maybe someone such as @jonorossi?

**Comment by thomaslevesque** [comment]
> I'm sad about the visibility, though.

Me too!

> Do you think it's worthwhile bothering a Castle.Core expert who might have insight? Maybe someone such as @jonorossi?

Sure!

**Comment by adamralph** [comment]
I think we should do the simplest thing possible to solve this issue, and that is, when calling `ToString()` for the purposes of generating a message, we catch errors and use the typename instead. We can examine more general solutions for other cases elsewhere.

**Comment by patrik-hagne** [comment]
Why not just disable the rule of strictness when writing the call
description? It could be a simple static bool in the strictness rule that
can be set from the write logic. Are there drawbacks of that? It could be
generalized into ICallRuleThatCanBeDisabled or something similar.

Den lör 18 feb. 2017 10:37Adam Ralph <notifications@github.com> skrev:

**Comment by blairconrad** [comment]
@adamralph, @patrik-hagne, your comments have merit, and I believe would help with the reported problem, but leave us open to errors in other situations, like the one I described in https://github.com/FakeItEasy/FakeItEasy/issues/457#issuecomment-73713213.

And maybe that's fine. Perhaps we could split that issue and decide whether or not to fix it, all the while adopting a "simple" fix for failure to _describe_ calls here.

However, I see that @thomaslevesque has already included a spec to cover the "equals" case I mention, and his solution deals with it.
Truth be told, I am a little leery about the described solution (I haven't looked at the code yet), as it does seem a little complicated. Being able to make the new interface `internal` would help with my worries a little. 

@thomaslevesque, I'm running out soon and haven't checked. How does your solution work with fake delegates? Or can they not even be strict (embarrassed that I don't know).

**Comment by jonorossi** [comment]
@blairconrad apologies I've read this long thread quite quickly, were you cc'ing me to ask:
- if `IFakeObject` is needed which I wouldn't think so, just replace the default `IProxyGenerationHook` which is `AllMethodsHook` which excludes `System.Object` virtuals for back compat with one that allows intercepting them. See the notes in the changelog under "3.0.0 beta 1" for more info.
- why `IFakeObject` needs to be public?

**Comment by blairconrad** [comment]
Apologies, @jonorossi. I should've been explicit. Thanks for showing up.

I was asking about why `IFakeObject` needs to be public.

*However*, I will also look at your notes about the `IProxyGenerationHook` to see if they help this situation.

**Comment by jonorossi** [comment]
> I was asking about why IFakeObject needs to be public.

I haven't looked at the pull request, however isn't this just the case where you would cause inconsistent accessibility with a public class (the proxy class) implementing an internal interface (`IFakeObject`). As far as I recall DynamicProxy doesn't support implementing interfaces explicitly.

**Comment by thomaslevesque** [comment]
> I think we should do the simplest thing possible to solve this issue, and that is, when calling `ToString()` for the purposes of generating a message, we catch errors and use the typename instead. We can examine more general solutions for other cases elsewhere.

I don't really agree with this approach. If we can find a solution that solves the general problem, why bother with a less than ideal solution for just part of the problem? The solution I suggested isn't as complex as it seems (I implemented it in an hour or two), and it works.

> Why not just disable the rule of strictness when writing the call
> description? It could be a simple static bool in the strictness rule that
> can be set from the write logic. Are there drawbacks of that? It could be
> generalized into `ICallRuleThatCanBeDisabled` or something similar.

That's an interesting idea, probably simpler than what I did. And we could do the same when we need to test equality between fakes (in fact, we could probably just use `ReferenceEquals` in this case).

> I haven't looked at the pull request, however isn't this just the case where you would cause inconsistent accessibility with a public class (the proxy class) implementing an internal interface (`IFakeObject`).

A public class can implement a non-public interface, it doesn't cause inconsistent accessibility. It's even a common pattern.

I never actually tried to implement an internal interface from another assembly with `InternalsVisibleTo`, but I expected it to work as well, so I was surprised to get an error when I tried to do it with Castle.Core.

> As far as I recall `DynamicProxy` doesn't support implementing interfaces explicitly.

That's not really an issue, as we can still detect if a method was called via the interface.

**Comment by jonorossi** [comment]
> A public class can implement a non-public interface, it doesn't cause inconsistent accessibility. It's even a common pattern.
> 
> I never actually tried to implement an internal interface from another assembly with InternalsVisibleTo, but I expected it to work as well, so I was surprised to get an error when I tried to do it with Castle.Core.

Apologies, yes of course, not sure what I thinking, it's late here. I do know that InternalsVisibleTo does work implementing an interface normally with the C# compiler, I've done that before. In that case we've probably got a bug in DynamicProxy.

**Comment by thomaslevesque** [comment]
> In that case we've probably got a bug in DynamicProxy.

Maybe. But according to the stack trace, the exception is thrown by `System.Reflection.Emit.TypeBuilder`, so maybe the bug is there, rather than in DynamicProxy... In any case, it would be great if there was a way around it.

**Comment by jonorossi** [comment]
@thomaslevesque what is the exception message?

**Comment by thomaslevesque** [comment]
I don't remember the exact message, but it roughly the same as in [this issue](https://github.com/castleproject/Core/issues/34#issue-21032365)

**Comment by thomaslevesque** [comment]
> But according to the stack trace, the exception is thrown by System.Reflection.Emit.TypeBuilder, so maybe the bug is there, rather than in DynamicProxy...

In fact, the problem is probably in DynamicProxy... I just made a quick test:

    class Program
    {
        static void Main(string[] args)
        {
            var assemblyBuilder = AppDomain.CurrentDomain.DefineDynamicAssembly(
                new AssemblyName("MyAssembly"),
                AssemblyBuilderAccess.RunAndSave);

            var moduleBuilder = assemblyBuilder.DefineDynamicModule("MyAssembly");

            var typeBuilder = moduleBuilder.DefineType(
                "Foo",
                TypeAttributes.Public | TypeAttributes.Class,
                typeof(object),
                new Type[] { typeof(IFoo) });

            var type = typeBuilder.CreateType();
            Console.WriteLine(type.AssemblyQualifiedName);
            Console.ReadLine();
        }
    }

    internal interface IFoo { }

This throws the error, but it works if I add `[assembly: InternalsVisibleTo("MyAssembly")]`. So Reflection.Emit _can_ implement an internal interface from another assembly, if the generated assembly is marked as friend.

**Comment by thomaslevesque** [comment]
In my test, it works if I don't specify the public key in `InternalsVisibleTo`:

    [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]

But fails if I specify it:

    [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")]

Also, it seems that `DynamicProxyGenAssembly2` has lost its strong name in Castle.Core 4.0.0; its `PublicKeyToken` is null... But I also tried with 3.3.3, where it does have a strong name, and it fails as well.

**Comment by blairconrad** [comment]
> In my test, it works if I don't specify the public key in `InternalsVisibleTo`

@thomaslevesque, I wonder if you have to omit the public key because your assembly is not strong-named. See castleproject/Core#33, or the discussion in castleproject/Core#30, which the former addresses.

**Comment by thomaslevesque** [comment]
> @thomaslevesque, I wonder if you have to omit the public key because your assembly is not strong-named. See castleproject/Core#33, or the discussion in castleproject/Core#30, which the former addresses.

Ah, I see. I thought I could always specify the public key for the friend assembly, even if the current assembly is not strong named. It's kind of weird that I need to exclude the public key in that case... Also the `InternalsVisibleToAttribute` documentation seems to say that the current and friend assembly either have to be both strong named or both not strong named. Does it mean that `DynamicProxyGenAssembly2` can be either strong-named or not? Or that `DynamicProxyGenAssembly2` and `DynamicProxyGenAssembly2, PublicKey=...` are actually different assemblies? I'm confused...

**Comment by thomaslevesque** [comment]
> Also, it seems that DynamicProxyGenAssembly2 has lost its strong name in Castle.Core 4.0.0; its PublicKeyToken is null...

Actually this might be because the mocked type comes from an assembly with no strong name

**Comment by adamralph** [comment]
Even if we do some trickery around strictness and `System.Object` methods, it there still a fundamental flaw here that we need to fix regardless? I.e. we call `ToString()` on objects when generating messages about them, and we don't catch exceptions on that call. Perhaps the object itself has a badly behaved `ToString` method, but we still want to generate the message, rather than have an unhandled exception.

**Comment by thomaslevesque** [comment]
> Even if we do some trickery around strictness and `System.Object` methods, is there still a fundamental flaw here that we need to fix regardless? I.e. we call `ToString()` on objects when generating messages about them, and we don't catch exceptions on that call. Perhaps the object itself has a badly behaved `ToString` method, but we still want to generate the message, rather than have an unhandled exception.

IMO we shouldn't call `ToString()` at all on a fake, whether it's strict or not. In my solution we call `IFakeObject.ToString()` instead. For non-fakes, I agree that we should catch the exception.

**Comment by jonorossi** [comment]
> Does it mean that DynamicProxyGenAssembly2 can be either strong-named or not? Or that DynamicProxyGenAssembly2 and DynamicProxyGenAssembly2, PublicKey=... are actually different assemblies? I'm confused...

Yes they are different assemblies, each `ModuleScope` has a strong and weak named dynamic assembly and will only be created when needed based on usage.

**Comment by blairconrad** [comment]
> 
IMO we shouldn't call `ToString()` at all on a fake, whether it's strict or not. In my solution we call `IFakeObject.ToString()` instead. For non-fakes, I agree that we should catch the exception.

ditto for `Equals` and `GetHashCode`, I assume.

**Comment by thomaslevesque** [comment]
> ditto for Equals and GetHashCode, I assume.

Yes

**Comment by adamralph** [comment]
From https://github.com/FakeItEasy/FakeItEasy/issues/457#issuecomment-280938007

> For non-fakes, I agree that we should catch the exception.

I suggest we release a very simple change now which catches the exception, falling back to the type name. That will solve the general case when the `ToString()` of _any_ object throws an exception, including this specific bug. After that, #984 may still have value, but we can pursue it for it's own merits independently of fixing this bug.

**Comment by thomaslevesque** [comment]
> I suggest we release a very simple change now which catches the exception, falling back to the type name. That will solve the general case when the ToString() of any object throws an exception, including this specific bug. After that, #984 may still have value, but we can pursue it for it's own merits independently of fixing this bug.

That makes sense.

As mentioned in your PR, I think we should prefix the type name with "Faked" if the value is a fake, since this is the default `ToString()` behavior for fakes.

**Comment by adamralph** [comment]
#993 actually fixes a more general bug, for any object that throws an exception in its ToString method, not just a strict fake. Is it worth altering this issue title and description to reflect that? Or should we raise another bug issue to reflect it?

**Comment by thomaslevesque** [comment]
I think altering the title is fine.

**Comment by blairconrad** [comment]
Agreed.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 3.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.1.0).

@Inspyro, look for your name in the release notes! :trophy:

**Comment by Inspyro** [comment]
@blairconrad Thanks :)
